### ISSUE TITLE
Broken behavior processing operations with defined 'output' class

### ISSUE THREAD

**Issue by Perf** [api-platform/core#2803]
Hello,
an exception is thrown when processing request by `WriteListener.php`.
At line 81 there is a call to get Iri from object of output class. But, obviously, it's not always possible, as the output class could be another object rather than ApiResource.
```php
if ($hasOutput) {
    $request->attributes->set('_api_write_item_iri', $this->iriConverter->getIriFromItem($controllerResult));
}
```
Exception:
> No item route associated with the type "Lexik\Bundle\JWTAuthenticationBundle\Response\JWTAuthenticationSuccessResponse" at /srv/api/vendor/api-platform/core/src/Bridge/Symfony/Routing/RouteNameResolver.php line 62

Here is a sample configuration of a custom operation:
```yaml
    collectionOperations:
        confirmation:
            method: 'POST'
            path: '/users/confirmation'
            access_control: 'is_anonymous()'
            messenger: 'input'
            input: 'App\Entity\User\Input\UserConfirm'
            output: 'Lexik\Bundle\JWTAuthenticationBundle\Response\JWTAuthenticationSuccessResponse'
```

**Some additional notes:**
There was a related issue #2550 that was fixed with #2628.
It worked for a while... and seems to be was 'broken' with: https://github.com/api-platform/core/commit/433ce07877655f245f164df36f1a249a7faf2abb#diff-3bc4e0ec1a3570f4deb961f3652800a6
So, now `$hasOutput` gets `true` at line 77, because now `$controllerResult` is an instance of `$outputMetadata['class']`
https://github.com/api-platform/core/blob/125dec90d05dc383e9214cd4968428323d0a1136/src/EventListener/WriteListener.php#L77
which it totally correct... but this leads to improper call of 
https://github.com/api-platform/core/blob/125dec90d05dc383e9214cd4968428323d0a1136/src/EventListener/WriteListener.php#L81

_Imho, at line 80 it's necessary to add additional check, if object in `$controllerResult` is an "ApiResource" object._

**Comment by dunglas** [comment]
@Perf good catch. A cleanest patch would be to pass the resource object related to the DTO to be able to populate this attribute. Would you want to work on a patch?

**Comment by Perf** [comment]
hey @dunglas,
yep, I could try to develop it. But need to clarify a few details.

- Why do we need to get an item Iri for a case when we respond with an object that is not (completely) related to any of Api Resources?

What we have in current config:
- `input` class is a kind of DTO that is related to api resource, but not really explicitly, as the only place that "links" this DTO with Api Resource is a yaml config of an Api Resource, where DTO class is stated as `input` class for a specific operation.
- `output` class is a completely not related object to any of api resources

So, as it seems, it should be possible to get an item Iri from `$resourceMetadata` and `$outputMetadata`.
https://github.com/api-platform/core/blob/125dec90d05dc383e9214cd4968428323d0a1136/src/EventListener/WriteListener.php#L75-L76
My assumption is correct?

**Comment by teohhanhui** [comment]
`_api_write_item_iri` is used for setting the correct HTTP response headers.

**Comment by teohhanhui** [comment]
Also, the `iri` in `ResourceMetadata` is not related to the item IRI. That's actually the IRI of the type, e.g. https://schema.org/Book

**Comment by Perf** [comment]
> `_api_write_item_iri` is used for setting the correct HTTP response headers: https://github.com/api-platform/core/blob/v2.4.3/src/EventListener/RespondListener.php#L65-L71
> 
> But yeah, I don't think we could make correct assumptions for all custom operations.

hey @teohhanhui 
thanks for clarification... and yes - here is a big question - if it's really mandatory to set `_api_write_item_iri`?
As for the subject case - it's seems to be absolutely nonsense.

**Comment by olivermack** [comment]
Actually I'm currently trying to get the head around this issue as well. I'm just curious why it should be absolutely mandatory to set the request attribute `_api_write_item_iri` at all - I mean, why not simply wrapping this portion in a try/catch?

An alternative could probably be to provide some sort of resource metadata which the `WriteListener` can utilize to determine if a result-iri should be obtained at all.

In my case I'm using the messenger to dispatch domain commands - in this very case handling GraphQL. The issue pops up as soon as you turn off the output (yah, I don't need to rely on synchronous processing with result here) - at least in the current `master` (did not compare with latest release `2.4.3`) the generated GraphQL schema is invalid because each Resource, even if it's a empty-response command DTO, needs to have `fields`... if not, GQL clients error with something like `Type MySampleCommand must define one or more fields` (via Apollo) or `MySampleCommand fields must be an object with field names as keys or a function which returns such an object.` (via GraphiQL).

**Comment by teohhanhui** [comment]
Should be fixed by #2910
